### PR TITLE
Reduce allocations on DiagnosticAnalyzerBase

### PR DESCRIPTION
**Pull request by paulomorgado**
Reduce allocations on DiagnosticAnalyzerBase

### CODE REVIEW

File: src/Microsoft.VisualStudio.Threading.Analyzers/DiagnosticAnalyzerBase.cs [hunk 1]
```diff
@@ -66,9 +66,23 @@ private static bool ConformsToAwaiterPattern(ITypeSymbol typeSymbol)
...
+            var hasOnCompletedMethod = false;
+            var hasIsCompletedProperty = false;
+
+            foreach (var member in typeSymbol.GetMembers())
+            {
+                hasGetResultMethod |= member.Name == nameof(TaskAwaiter.GetResult) && member is IMethodSymbol m && m.Parameters.IsEmpty;
+                hasOnCompletedMethod |= member.Name == nameof(TaskAwaiter.OnCompleted) && member is IMethodSymbol;
+                hasIsCompletedProperty |= member.Name == nameof(TaskAwaiter.IsCompleted) && member is IPropertySymbol;
+
+                if (hasGetResultMethod || hasOnCompletedMethod || hasIsCompletedProperty)
```
Comment (Therzok) [inline src/Microsoft.VisualStudio.Threading.Analyzers/DiagnosticAnalyzerBase.cs]: The previous code would check for all 3 constraints to be satisfied simultaneously.

The new one checks only for one of them to be satisfied. Was this intended? Should it not be:
```
if (hasGetResultMethod && hasOnCompletedMethod && hasIsCompletedProperty)
```

Comment (AArnott) [reply src/Microsoft.VisualStudio.Threading.Analyzers/DiagnosticAnalyzerBase.cs]: Ouch. You're absolutely right. I'll send a follow-up PR to fix this.

Comment (AArnott) [reply src/Microsoft.VisualStudio.Threading.Analyzers/DiagnosticAnalyzerBase.cs]: #495

Comment (paulomorgado) [reply src/Microsoft.VisualStudio.Threading.Analyzers/DiagnosticAnalyzerBase.cs]: Sorry about that! :disappointed:

**Pull request merged**